### PR TITLE
Enhancement: Add rector config file for easy migration

### PR DESCRIPTION
**Pull request by OskarStark**
This PR is based on the work of https://github.com/thecodingmachine/safe

I have a lot of projects where I want to run a continuous automated refactoring for **properties** to **methods**, so I ended up repeating the following code over and over again:
```php
$rectorConfig->ruleWithConfiguration(
    PropertyFetchToMethodCallRector::class,
    [
        new PropertyFetchToMethodCall(Generator::class, 'boolean', 'boolean'),
        new PropertyFetchToMethodCall(Generator::class, 'city', 'city'),
        new PropertyFetchToMethodCall(Generator::class, 'dateTime', 'dateTime'),
        new PropertyFetchToMethodCall(Generator::class, 'dateTimeThisYear', 'dateTimeThisYear'),
        new PropertyFetchToMethodCall(Generator::class, 'email', 'email'),
        new PropertyFetchToMethodCall(Generator::class, 'firstName', 'firstName'),
        new PropertyFetchToMethodCall(Generator::class, 'firstNameMale', 'firstNameMale'),
        new PropertyFetchToMethodCall(Generator::class, 'lastName', 'lastName'),
        new PropertyFetchToMethodCall(Generator::class, 'name', 'name'),
        new PropertyFetchToMethodCall(Generator::class, 'password', 'password'),
        new PropertyFetchToMethodCall(Generator::class, 'phoneNumber', 'phoneNumber'),
        new PropertyFetchToMethodCall(Generator::class, 'sentence', 'sentence'),
        new PropertyFetchToMethodCall(Generator::class, 'sha256', 'sha256'),
        new PropertyFetchToMethodCall(Generator::class, 'slug', 'slug'),
        new PropertyFetchToMethodCall(Generator::class, 'text', 'text'),
        new PropertyFetchToMethodCall(Generator::class, 'url', 'url'),
        new PropertyFetchToMethodCall(Generator::class, 'word', 'word'),
        new PropertyFetchToMethodCall(Generator::class, 'words', 'words'),
    ]);
```

I think it is better to provide this file via the vendor itself.
Check the README for usage details

### Todos
* [x] complete the list of deprecated properties

WDYT?

friendly ping @silasjoisten @mvhirsch @localheinz

### CODE REVIEW

File: rector-migrate.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+declare(strict_types=1);
+
+use Rector\Config\RectorConfig;
+use Rector\Transform\Rector\Assign\PropertyFetchToMethodCallRector;
+use Rector\Transform\ValueObject\PropertyFetchToMethodCall;
+
+// This file configures rector/rector to replace all PHP functions with their equivalent "safe" functions.
```
Comment (OskarStark) [inline rector-migrate.php:9]:
```suggestion
// This file configures rector/rector to replace all deprecated property usages with their equivalent functions.
```

File: README.md [hunk 2]
```diff
@@ -57,6 +57,37 @@ for ($i = 0; $i < 3; $i++) {
...
+
+to run `rector/rector`.
+
+*Note:* do not forget to replace "src/" with the path to your source directory.
+
+Another way is to use it in your `rector.php` file:
+
+```php
+    $faker = require 'vendor/fakerphp/faker/rector-migrate.php';
+    $faker($rectorConfig);
```
Comment (OskarStark) [inline README.md:88]:
```suggestion
$faker = require 'vendor/fakerphp/faker/rector-migrate.php';
$faker($rectorConfig);
```

**Comment by pimjansen** [comment]
I dont think this should be part of Faker itself. It would mean Faker has a locked dependency on Rector which is not what you want. 

Just publish it as separate package? I think this would be the right thing to go here though

**Comment by pierredup** [comment]
> It would mean Faker has a locked dependency on Rector which is not what you want.

There is no locked dependency, it is an optional extra file. If you use Rector in your project, you can use this additional config file, but if you don't use Rector, it won't be part of any of your dependencies

File: README.md [hunk 3]
```diff
@@ -57,6 +57,37 @@ for ($i = 0; $i < 3; $i++) {
...
+```
+
+to run `rector/rector`.
+
+*Note:* do not forget to replace "src/" with the path to your source directory.
+
+Another way is to use it in your `rector.php` file:
+
+```php
+$faker = require 'vendor/fakerphp/faker/rector-migrate.php';
```
Comment (pierredup) [inline README.md:87]: This can use the import from Rector

```suggestion
$rectorConfig->import('vendor/fakerphp/faker/rector-migrate.php');
```

**Comment by OskarStark** [comment]
> There is no locked dependency, it is an optional extra file. If you use Rector in your project, you can use this additional config file, but if you don't use Rector, it won't be part of any of your dependencies

Exactly 👍

File: README.md (hunk 3 above)
Comment (OskarStark) [reply README.md:87]: Oh I didn't know about this 😄 

Thanks!

File: rector-migrate.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        'sha256',
+        'slug',
+        'text',
+        'url',
+        'word',
+        'words',
+    ];
+
+    $methodCalls = [];
+    foreach ($properties as $property) {
```
Comment (localheinz) [inline rector-migrate.php:33]:
```suggestion
    $methodCalls = [];

    foreach ($properties as $property) {
```

This should fix the broken build.

File: rector-migrate.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+declare(strict_types=1);
+
+use Rector\Config\RectorConfig;
+use Rector\Transform\Rector\Assign\PropertyFetchToMethodCallRector;
+use Rector\Transform\ValueObject\PropertyFetchToMethodCall;
+
+// This file configures rector/rector to replace all deprecated property usages with their equivalent functions.
+return static function (RectorConfig $rectorConfig): void {
```
Comment (localheinz) [inline rector-migrate.php:10]: @OskarStark 

How do you feel about adding a workflow for demonstrating that this configuration works?

Perhaps https://github.com/OskarStark/Faker/pull/1?

File: rector-migrate.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    $methodCalls = [];
+    foreach ($properties as $property) {
+        $methodCalls[] = new PropertyFetchToMethodCall(Generator::class, $property, $property);
+    }
+
+    $rectorConfig->ruleWithConfiguration(
+        PropertyFetchToMethodCallRector::class,
+        $methodCalls
+    );
```
Comment (localheinz) [inline rector-migrate.php:40]:
```suggestion
    ];

    $rectorConfig->ruleWithConfiguration(
        PropertyFetchToMethodCallRector::class,
        array_map(static function (string $property): PropertyFetchToMethodCall {
            return new PropertyFetchToMethodCall(Generator::class, $property, $property);
        }, $properties)
    );
```

File: rector-migrate.php (hunk 5 above)
Comment (OskarStark) [reply rector-migrate.php:10]: Make sense

**Comment by OskarStark** [comment]
Can someone please approve the new workflow to run? Thanks

**Comment by bram-pkg** [comment]
Should be running now.

File: rector-migrate.php [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            throw new \InvalidArgumentException(sprintf('Property "%s" does not exist in class "%s"', $property, \Faker\Generator::class));
+        }
+
+        $methodCalls[] = new PropertyFetchToMethodCall(Generator::class, $property, $property);
+    }
+
+    $rectorConfig->ruleWithConfiguration(
+        PropertyFetchToMethodCallRector::class,
+        $methodCalls
+    );
```
Comment (localheinz) [inline rector-migrate.php:164]:
```suggestion
use Faker\Generator;
use PHPStan\PhpDocParser;
use Rector\Config\RectorConfig;
use Rector\Transform\Rector\Assign\PropertyFetchToMethodCallRector;
use Rector\Transform\ValueObject\PropertyFetchToMethodCall;

// This file configures rector/rector to replace all deprecated property usages with their equivalent functions.
return static function (RectorConfig $rectorConfig): void {
    $reflection = new \ReflectionClass(Generator::class);

    $docBlock = $reflection->getDocComment();

    $lexer = new PhpDocParser\Lexer\Lexer();

    $constExprParser = new PhpDocParser\Parser\ConstExprParser();

    $phpDocParser = new PhpDocParser\Parser\PhpDocParser(
        new PhpDocParser\Parser\TypeParser($constExprParser),
        $constExprParser
    );

    $phpDocNode = $phpDocParser->parse(new PhpDocParser\Parser\TokenIterator($lexer->tokenize($docBlock)));

    $properties = array_map(static function (PhpDocParser\Ast\PhpDoc\PropertyTagValueNode $propertyTagValueNode): string {
        if (!preg_match('/^\$(?P<name>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)$/', $propertyTagValueNode->propertyName, $matches)) {
            throw new \RuntimeException(\sprintf(
                'Property name "%s" is not valid.',
                $propertyTagValueNode->propertyName
            ));
        }

        return $matches['name'];
    }, $phpDocNode->getPropertyTagValues());

    $rectorConfig->ruleWithConfiguration(
        PropertyFetchToMethodCallRector::class,
        array_map(static function (string $property): array {
            return new PropertyFetchToMethodCall(Generator::class, $property, $property);
        }, $properties)
    );
```

Perhaps this makes sense after running

```php
composer require phpstan/phpdoc-parser:^1.14.0 --working-dir=vendor-bin/rector
```

?

Comment (OskarStark) [reply rector-migrate.php:164]: In this case you can only use this config if you have PHPStan running... So I am more for the static version to be honest.

Anyway, all properties are already deprecated and I would not overengineer it.

File: rector-migrate.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        'words',
+        'year',
+    ];
+
+    $methodCalls = [];
+
+    foreach ($properties as $property) {
+        if (!property_exists(\Faker\Generator::class, $property)) {
+            throw new \InvalidArgumentException(sprintf('Property "%s" does not exist in class "%s"', $property, \Faker\Generator::class));
+        }
```
[the remark this answers is not in the document]
Comment (OskarStark) [reply rector-migrate.php:156]: Hmm looks like this piece of code is not working as I expected it:
https://3v4l.org/GZbT5

Comment (OskarStark) [inline rector-migrate.php:156]:
```suggestion
```

File: rector-migrate.php [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        'windowsPlatformToken',
+        'word',
+        'words',
+        'year',
+    ];
+
+    $methodCalls = [];
+
+    foreach ($properties as $property) {
+
```
Comment (OskarStark) [inline rector-migrate.php:154]:
```suggestion
```

**Review by localheinz — APPROVED** [review]
👍

**Comment by localheinz** [comment]
Thank you, @bram-pkg, @OskarStark, @pierredup, and @pimjansen!

**Comment by DvDty** [comment]
What is the point of running rector on each PR, if the report is ignored?

**Pull request merged**